casanovo/denovo/model.py

```diff
@@ -176,12 +193,592 @@ def forward(
+
+        Parameters
+        ----------
+        spectra : torch.Tensor of shape (n_spectra, n_peaks, 2)
+            The spectra to embed. Axis 0 represents a mass spectrum, axis 1
+            contains the peaks in the mass spectrum, and axis 2 is essentially
+            a 2-tuple specifying the m/z-intensity pair for each peak. These
+            should be zero-padded, such that all of the spectra in the batch
+            are the same length.
+        precursors : torch.Tensor of size (n_spectra, 3)
+            The measured precursor mass (axis 0), precursor charge (axis 1), and
+            precursor m/z (axis 2) of each MS/MS spectrum.
+
+        Returns
+        -------
+        tokens : torch.Tensor of shape
+        (n_spectra, max_length, n_amino_acids, n_beams)
+            The token sequence for each spectrum.
+        scores : torch.Tensor of shape
+        (n_spectra, max_length, n_amino_acids, n_beams)
+            The score for each amino acid.
+        """
+        memories, mem_masks = self.encoder(spectra)
+
+        # Sizes:
+        batch = spectra.shape[0]  # B
+        length = self.max_length + 1  # L
+        vocab = self.decoder.vocab_size + 1  # V
+        beam = self.n_beams  # S
+
+        # Initialize scores and tokens:
+        scores = torch.zeros(batch, length, vocab, beam)
+        scores = scores.type_as(spectra)
+        scores[scores == 0] = torch.nan
+        tokens = torch.zeros(batch, length, beam)
+        tokens = tokens.type_as(spectra).long()
+
+        # Create cache for decoded beams
+        (
+            cache_scores,
+            cache_tokens,
+            cache_idx_dict,
+            cache_seq_dict,
+            cache_score_dict,
+        ) = self._create_beamsearch_cache(scores, tokens)
+        # Create the first prediction:
+        pred, _ = self.decoder(None, precursors, memories, mem_masks)
+        _, idx = torch.topk(pred[:, 0, :], beam, dim=1)
+        tokens[:, 0, :] = idx
+        scores[:, :1, :, :] = einops.repeat(pred, "B L V -> B L V S", S=beam)
+
+        # Make precursors and memories the right shape for decoding:
+        precursors = einops.repeat(precursors, "B L -> (B S) L", S=beam)
+        mem_masks = einops.repeat(mem_masks, "B L -> (B S) L", S=beam)
+        memories = einops.repeat(memories, "B L V -> (B S) L V", S=beam)
+
+        # The main decoding loop:
+        for idx in range(1, self.max_length + 1):
+            scores = einops.rearrange(scores, "B L V S -> (B S) L V")
+            tokens = einops.rearrange(tokens, "B L S -> (B S) L")
+
+            # Terminate beams exceeding precursor m/z tolerance and track all terminated beams
+            finished_beams_idx, tokens = self._terminate_finished_beams(
+                tokens, precursors, idx
+            )
+
+            # Cache terminated beams, group and order by fitting precursor m/z and confidence score
+            (
```

Comment:
**suggestion:** `cache_idx_dict`, `cache_seq_dict`, and `cache_score_dict` are passed by reference in the function, so you don't necessarily need to return them. I'd rewrite this function without a return value to simplify the code a bit.